portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -21,25 +21,81 @@
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList;
 import com.google.gson.*;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  ArrayList<String> comments = new ArrayList<String>();
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numComments;
+    try {
+      numComments = Integer.parseInt(request.getParameter("number-comments"));
+    } catch(Exception e) {
+      numComments = 5;
+      System.out.println(request.getParameter("number-comments"));
+      System.out.println(request.getParameter("authuser"));
+      System.out.println(request.getQueryString());
+    }
+    
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
```

Comment:
Consider using FetchOptions [1] to be more efficient. Suppose you have 10k comments and only wish to display 5. Retrieving all of them is wasteful.
[1] https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions

Reply:
Correcting myself: per page 132 of [1] Iterables only retrieve a small batch of entries as needed as you advance, so efficiency won't be too much of a problem. But using FetchOptions can still save you a loop :)

[1] http://netoearth.com/book/ProgrammingGoogleAppEngine.pdf

Reply:
Thank you for making me aware of this. I had been searching for a way to limit the size of the query but it looks like my mistake was trying to add a filter when creating it as opposed to limiting the size of the response when executing. I will look into changing this to make it more efficient